Reintroduce TypeScript docs (#2153)
wasp-lang/wasp: web/docs/data-model/operations/queries.md

- `GetFilteredTasks` is also a generic type automatically generated by Wasp, based on the Query declaration for `getFilteredTasks`.

Use these types to type the Query's implementation.
Annotating Queries is optional, but very helpful.
Doing so properly types the Query's context.

In this case, TypeScript will know the `context.entities` object must include the `Task` entity.
TypeScript also knows whether the `context` object includes user information (it depends on whether your Query uses auth).

The generated types are generic and accept two optional type arguments: `Input` and `Output`.

1. `Input` - The argument (i.e., payload) received by the Query function.
2. `Output` - The Query function's return type.

Use these type arguments to type the Query's inputs and outputs.

For example, the above code says that the Query `getTasks` doesn't expect any arguments (its input type is `void`), but it does return a list of tasks (its output type is `Task[]`).
On the other hand, the Query `getFilteredTasks` expects an object of type `{ isDone: boolean }`. This type is derived from the `Task` entity type.

> Suppose you don't care about typing the Query's inputs and outputs. In that case, you can omit both type arguments, and TypeScript will infer the most general types (i.e., `never` for the input, `unknown` for the output.).

Reviewer: Maybe simplify the language a bit:
```suggestion
If you don't care about typing the Query's inputs and outputs, you can omit both type arguments. TypeScript will then infer the most general types (`never` for the input and `unknown` for the output.).
```

Even though specifying `Input` or `Output` is completely optional, we highly recommend you spcify them. Doing so gives you:
  - Type support for the arguments and the return value inside the implementation.
  - **Full-stack type safety**. We'll explore what this means when we discuss calling the Query from the client.

Read more about type support for implementing Queries in the [API Reference](#implementing-queries).

> :::tip Inferring the return type

Author: It's annoying that I have to repeat some of these admonitions twice because they slightly differ (it's usually only mentions of Queries or Actions).

Still, duplication seemed better than coming up with a component that interpolates props into the text.

If don't want to explicitly type the Query's return value, the `satisfies` keyword tells TypeScript to infer it automatically:
```typescript
const getFoo = (async (_args, context) => {
  const foos = await context.entities.Foo.findMany()
  return {
    foos,
    message: "Here are some foos!",
    queriedAt: new Date(),
  }
}) satisfies GetFoo
```
From the snippet above, TypeScript knows:
1. The correct type for `context`. 
2. The Query's return type is `{ foos: Foo[], message: string, queriedAt: Date }`.

> If you don't need the context, you can skip specifying the Query's type (and arguments):

Reviewer: This feels cool, but I'm on the fence if this will ever happen in real-world project?

Author: Probably not, but I thought I'd include it to show that the arguments are optional. You think it's better if I remove it?

Reviewer: If there is a chance of users shooting themselves in the foot because of this, I'd remove it. Otherwise, it's fine by me.

Author: Haha, nice uno reverse.

I'm keeping it then, I don't think it's a problem.
